File: dailyQuestion/2020-07/07-04/python/solution_dp.py
def longestValidParentheses(s: str) -> int:
    result = 0
    if not s:
        return result
    n = len(s)
    dp = [0] * n
    for i in range(1, n):
        if s[i] == ")":
            if s[i - 1] == "(":
                dp[i] = dp[i - 2] + 2
            elif i - dp[i - 1] > 0 and s[i - dp[i - 1] - 1] == "(":
                if i - dp[i - i] >= 2:
                    dp[i] = dp[i - 1] + dp[i - dp[i - 1] - 2] + 2
                else:
                    dp[i] = dp[i - 1] + 2
            result = max(result, dp[i])
    return result


if __name__ == "__main__":
    s = "((()))()"
    result = longestValidParentheses(s)
    print(result)

File: dailyQuestion/2020-07/07-04/golang/solution_dp.go
package main

import "fmt"

func longestValidParentheses(s string) int {
	n := len(s)
	result := 0
	if n == 0 {
		return result
	}
	dp := make([]int, len(s))
	for i := 1; i < n; i++ {
		if s[i] == ')' {
			if s[i-1] == '(' {
				if i >= 2 {
					dp[i] = dp[i-2] + 2
				} else {
					dp[i] = 2
				}
			} else if i-dp[i-1] > 0 && s[i-dp[i-1]-1] == '(' {
				if i-dp[i-1] >= 2 {
					dp[i] = dp[i-1] + dp[i-dp[i-1]-2] + 2
				} else {
					dp[i] = dp[i-1] + 2
				}
			}
			result = max(result, dp[i])
		}
	}
	return result
}

func max(x, y int) int {
	if x > y {
		return x
	}
	return y
}

func main() {
	s := "(())()"
	result := longestValidParentheses(s)
	fmt.Println(result)
}
